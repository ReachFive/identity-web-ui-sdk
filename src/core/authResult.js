import { parseJwtTokenPayload } from '../helpers/jwt'
import { logError } from '../helpers/logger'

export function enrichAuthResult(response) {
    if (response.idToken) {
        try {
            const idTokenPayload = parseJwtTokenPayload(response.idToken);

            return {
                ...response,
                idTokenPayload
            };
        } catch (e) {
            logError('id token parsing error: ' + e);
        }
    }

    return response;
}

export function isAuthResult(value) {
    return value && (value.accessToken || value.idToken);
}

export function isErrorResponse(value) {
    return value && value.error;
}
