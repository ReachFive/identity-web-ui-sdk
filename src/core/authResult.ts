import { AuthResult, ErrorResponse } from '@reachfive/identity-core';

import { parseJwtTokenPayload } from '../helpers/jwt';
import { logError } from '../helpers/logger';

export function enrichAuthResult(response: AuthResult) {
    if (response.idToken) {
        try {
            const idTokenPayload = parseJwtTokenPayload(response.idToken);

            return {
                ...response,
                idTokenPayload,
            };
        } catch (e) {
            logError('id token parsing error: ' + e);
        }
    }

    return response;
}

export function isAuthResult(value: unknown): value is AuthResult {
    return (
        value !== null &&
        typeof value === 'object' &&
        ('accessToken' in value || 'idToken' in value)
    );
}

export function isErrorResponse(value: unknown): value is ErrorResponse {
    return value !== null && typeof value === 'object' && 'error' in value;
}
