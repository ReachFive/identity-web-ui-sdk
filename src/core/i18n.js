import isPlainObject from 'lodash-es/isPlainObject';

export default function resolveI18n(defaultMessages, messages = {}) {
    const mergedMessages = {
        ...defaultMessages,
        ...flattenObject(messages)
    };

    return (key, params) => {
        const template = mergedMessages[key] || key;

        return params
            ? Object.keys(params).reduce((acc, param) => acc.replace(`{${param}}`, params[param]), template)
            : template;
    };
}

function flattenObject(object, prefix = []) {
    return isPlainObject(object)
        ? Object.keys(object).reduce((acc, key) => ({
            ...acc,
            ...flattenObject(object[key], [...prefix, key])
        }), {})
        : ({ [prefix.join('.')]: object });
}
