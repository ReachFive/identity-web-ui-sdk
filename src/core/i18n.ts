export type I18nMessages = Record<string, string>

export type I18nNestedMessages = Record<string, string | I18nMessages>

export type I18nMessageParams = Record<string, unknown>

export type I18nResolver = (key: string, params?: I18nMessageParams, fallback?: (params?: I18nMessageParams) => string) => string

export function resolveI18n<T extends (Record<string, T> | string)>(defaultMessages: I18nMessages = {}, messages: I18nNestedMessages = {}): I18nResolver {
    const mergedMessages: I18nMessages = {
        ...defaultMessages,
        ...flattenObject(messages)
    };

    return (key: string, params?: I18nMessageParams, fallback?: (params?: I18nMessageParams) => string) => {
        const template = mergedMessages[key] ?? fallback?.(params) ?? key;

        return params
            ? Object.keys(params).reduce((acc, param) => acc.replace(`{${param}}`, params[param] as string), template)
            : template;
    };
}

export default resolveI18n

function flattenObject(object: I18nNestedMessages | string, prefix: string[] = []): Record<string, string> {
    return typeof object === "object" && object !== null
        ? Object.keys(object).reduce((acc, key) => ({
            ...acc,
            ...flattenObject(object[key], [...prefix, key])
        }), {})
        : ({ [prefix.join('.')]: object });
}
