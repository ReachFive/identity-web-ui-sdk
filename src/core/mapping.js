import { getValue, setValue } from '../helpers/propertyHelpers';

export class PathMapping {
    constructor(modelPath) {
        this.modelPath = modelPath;
    }

    bind(model) {
        return getValue(model, this.modelPath);
    }

    unbind(model, value) {
        return setValue(model, this.modelPath, value);
    }
}
