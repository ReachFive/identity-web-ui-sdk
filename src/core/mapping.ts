import { ProfileAddress } from '@reachfive/identity-core';
import { getValue, setValue } from '../helpers/propertyHelpers';

export interface PathMapping {
    bind<T extends Record<string, unknown>>(model: T): unknown;
    unbind<T extends Record<string, unknown>, V>(model: T, value: V): T | V;
}

export class DefaultPathMapping implements PathMapping {
    protected readonly modelPath: string;

    constructor(modelPath: string) {
        this.modelPath = modelPath;
    }

    bind<T extends Record<string, unknown>>(model: T) {
        return getValue(model, this.modelPath);
    }

    unbind<T extends Record<string, unknown>, V>(model: T, value: V) {
        return setValue(model, this.modelPath, value);
    }
}

export class AddressPathMapping implements PathMapping {
    protected readonly property: string;

    constructor(property: string) {
        this.property = property;
    }

    private defaultIndex(model: { addresses?: ProfileAddress[] }) {
        const defaultIndex = (model.addresses ?? []).findIndex(address => address.isDefault);
        return defaultIndex > -1 ? defaultIndex : 0;
    }

    bind<T extends { addresses?: ProfileAddress[] }>(model: T) {
        return getValue(model, `addresses.${this.defaultIndex(model)}.${this.property}`);
    }

    unbind<T extends { addresses?: ProfileAddress[] }, V>(model: T, value: V) {
        return setValue(model, `addresses.${this.defaultIndex(model)}.${this.property}`, value);
    }
}
