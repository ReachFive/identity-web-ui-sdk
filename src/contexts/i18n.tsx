import React, { ComponentType, PropsWithChildren, useMemo } from 'react';

import { I18nMessages, I18nNestedMessages, I18nResolver, resolveI18n } from '../core/i18n';

export interface Props {
  defaultMessages?: I18nMessages
  messages?: I18nNestedMessages
}

export const I18nContext = React.createContext<I18nResolver | undefined>(undefined);

export function useI18n(): I18nResolver {
  const context = React.useContext(I18nContext);
  if (!context) {
    throw new Error('No I18nContext provided');
  }

  return context;
}

export interface I18nProps {
  i18n: I18nResolver
}

export type WithI18n<P> = P & I18nProps

export function withI18n<T extends I18nProps = I18nProps>(WrappedComponent: ComponentType<T>) {
  const displayName = WrappedComponent.displayName || WrappedComponent.name || "Component";

  const ComponentWithI18n = (props: Omit<T, keyof I18nProps>) => {
    const i18n = useI18n();
    return <WrappedComponent {...({ i18n })} {...(props as T)} />;
  };

  ComponentWithI18n.displayName = `withI18n(${displayName})`;

  return ComponentWithI18n;
}

export function I18nProvider({ children, defaultMessages, messages }: PropsWithChildren<Props>): JSX.Element | null {
  const resolver = useMemo(() => resolveI18n(defaultMessages, messages), [defaultMessages, messages])
  return <I18nContext.Provider value={resolver}>{children}</I18nContext.Provider>
}
