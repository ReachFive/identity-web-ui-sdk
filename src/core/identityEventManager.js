import { enrichAuthResult } from './authResult'
import EventManager from '../helpers/eventManager'

export function createEventManager() {
    const eventManager = new EventManager();

    return {
        on(eventName, listener) {
            eventManager.on(eventName, listener);
        },

        off(eventName, listener) {
            eventManager.off(eventName, listener)
        },

        fireEvent(eventName, data) {
            if (eventName === 'authenticated') {
                eventManager.fire(eventName, enrichAuthResult(data));
            } else {
                eventManager.fire(eventName, data);
            }
        }
    }
}
