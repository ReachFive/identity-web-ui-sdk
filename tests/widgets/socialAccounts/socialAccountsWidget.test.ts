/**
 * @jest-environment jest-fixed-jsdom
 */

import { beforeEach, describe, expect, jest, test } from '@jest/globals';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event'
import '@testing-library/jest-dom/jest-globals'
import 'jest-styled-components';

import { type Profile, type Client } from '@reachfive/identity-core';

import { type I18nMessages } from '../../../src/core/i18n';
import type { Config } from '../../../src/types';

import SocialAccountsWidget from '../../../src/widgets/socialAccounts/socialAccountsWidget';
import { componentGenerator, snapshotGenerator } from '../renderer';
import { ComponentProps } from 'react';

const defaultI18n: I18nMessages = {}

describe('Snapshot', () => {
    const generateSnapshot = (
        options: ComponentProps<typeof SocialAccountsWidget>,
        config: Partial<Config> = {},
        socialIdentities: Profile['socialIdentities'] = []
    ) => {
        // @ts-expect-error partial Client
        const apiClient: Client = {
            // @ts-expect-error partial Profile
            getUser: jest.fn<Client['getUser']>().mockResolvedValue({ socialIdentities }),
            unlink: jest.fn<Client['unlink']>(),
            on: jest.fn(),
            off: jest.fn(),
        };

        const generate = snapshotGenerator(SocialAccountsWidget, apiClient, defaultI18n)

        return generate(options, config)
    };

    test('basic', generateSnapshot(
        { accessToken: 'azerty' },
        {},
        [
            { id: '123456778', provider: 'facebook', username: 'John Doe' }
        ]
    ))
})

describe('DOM testing', () => {
    const getUser = jest.fn<Client['getUser']>()
    const unlink = jest.fn<Client['unlink']>()
    const on = jest.fn()
    const off = jest.fn()

    const onError = jest.fn()
    const onSuccess = jest.fn()
    
    beforeEach(() => {
        getUser.mockClear()
        unlink.mockClear()
        on.mockClear()
        off.mockClear()
        onError.mockClear()
        onSuccess.mockClear()
    })

    const generateComponent = async (
        options: ComponentProps<typeof SocialAccountsWidget>,
        config: Partial<Config> = {},
        socialIdentities: Profile['socialIdentities'] = []
    ) => {
        // @ts-expect-error partial Client
        const apiClient: Client = {
            // @ts-expect-error partial Profile
            getUser: getUser.mockResolvedValue({ socialIdentities }),
            unlink,
            on,
            off,
        };

        const generate = componentGenerator(SocialAccountsWidget, apiClient, defaultI18n)
        
        return await generate(options, config)
    }

    describe('with default config', () => {
        test('no identity', async () => {
            expect.assertions(6);

            await generateComponent({ accessToken: 'azerty', onError, onSuccess });

            expect(screen.queryByText('socialAccounts.noLinkedAccount')).toBeInTheDocument();
            expect(screen.queryByText('socialAccounts.linkNewAccount')).toBeInTheDocument();
        })

        test('with existing identity', async () => {
            expect.assertions(7);

            await generateComponent(
                { accessToken: 'azerty', onError, onSuccess },
                {},
                [
                    { id: '123456789', provider: 'facebook', username: 'John Doe' },
                ]
            );

            expect(screen.queryByText('socialAccounts.noLinkedAccount')).not.toBeInTheDocument();
            expect(screen.queryByTestId('identity-facebook')).toBeInTheDocument();
            expect(screen.queryByText('socialAccounts.linkNewAccount')).toBeInTheDocument();
        })

        test('with all identities configured', async () => {
            expect.assertions(9);

            await generateComponent(
                { accessToken: 'azerty', onError, onSuccess },
                {},
                [
                    { id: '123456789', provider: 'facebook', username: 'John Doe' },
                    { id: '987654321', provider: 'google', username: 'John Doe' },
                    { id: '000000000', provider: 'line', username: 'John Doe' },
                ]
            );

            expect(screen.queryByText('socialAccounts.noLinkedAccount')).not.toBeInTheDocument();
            expect(screen.queryByTestId('identity-facebook')).toBeInTheDocument();
            expect(screen.queryByTestId('identity-google')).toBeInTheDocument();
            expect(screen.queryByTestId('identity-line')).toBeInTheDocument();
            expect(screen.queryByText('socialAccounts.linkNewAccount')).not.toBeInTheDocument();
        })

        test('unlink identity', async () => {
            const user = userEvent.setup()

            unlink.mockResolvedValue()

            await generateComponent(
                { accessToken: 'azerty', onError, onSuccess },
                {},
                [
                    { id: '123456789', provider: 'facebook', username: 'John Doe' },
                    { id: '987654321', provider: 'google', username: 'John Doe' },
                    { id: '000000000', provider: 'line', username: 'John Doe' },
                ]
            );

            expect(screen.queryByTestId('identity-google')).toBeInTheDocument();
            
            const unlinkBtn = screen.getByTestId('identity-google-unlink');
            expect(unlinkBtn).toBeInTheDocument();
            
            await user.click(unlinkBtn)

            expect(unlink).toHaveBeenCalledWith(
                expect.objectContaining({
                    accessToken: 'azerty',
                    identityId: '987654321' // Google
                })
            )

            expect(screen.queryByTestId('identity-google')).not.toBeInTheDocument();

            expect(onSuccess).toBeCalled()
            expect(onError).not.toBeCalled()
        })

        test('unlink api failure', async () => {
            const user = userEvent.setup()

            unlink.mockRejectedValue(new Error('Unexpected error'))

            await generateComponent(
                { accessToken: 'azerty', onError, onSuccess },
                {},
                [
                    { id: '123456789', provider: 'facebook', username: 'John Doe' },
                    { id: '987654321', provider: 'google', username: 'John Doe' },
                    { id: '000000000', provider: 'line', username: 'John Doe' },
                ]
            );

            const unlinkBtn = screen.getByTestId('identity-google-unlink');
            await user.click(unlinkBtn)

            expect(unlink).toBeCalled()

            expect(onSuccess).not.toBeCalled()
            expect(onError).toBeCalled()
        })
    })
})
