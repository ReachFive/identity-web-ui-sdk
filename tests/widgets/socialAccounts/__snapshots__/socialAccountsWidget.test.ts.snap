// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot basic 1`] = `
.c1 {
  text-align: center;
  margin-bottom: 10px;
}

.c2 {
  color: #495057;
}

.c0 {
  font-size: 14px;
  -webkit-transition: -webkit-transform 400ms ease, opacity 400ms ease;
  -webkit-transition: transform 400ms ease, opacity 400ms ease;
  transition: transform 400ms ease, opacity 400ms ease;
  opacity: 0;
  padding: 20px;
  border-radius: 3px;
  background-color: #ffffff;
  max-width: 400px;
  box-sizing: border-box;
  margin: 0 auto;
}

.c4 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  text-align: center;
  font-weight: bold;
  vertical-align: middle;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  touch-action: manipulation;
  cursor: pointer;
  color: #495057;
  background-color: #ffffff;
  border: 1px solid #ced4da;
  padding: 9px 12px;
  font-size: 14px;
  line-height: 1.428571429;
  border-radius: 3px;
  -webkit-transition: all .15s ease-in-out;
  transition: all .15s ease-in-out;
}

.c4:focus {
  outline: 0;
  box-shadow: 0 0 0 3px rgba(206,212,218,0.5);
}

.c4:hover,
.c4:active {
  color: #495057;
  background-color: #ebebeb;
  border-color: #b7c0c8;
}

.c4[disabled] {
  opacity: .65;
}

.c3 {
  margin-top: 10px;
}

<div>
  <div
    class="c0"
  >
    <div>
      <div
        class="c1 c2"
      >
        socialAccounts.noLinkedAccount
      </div>
    </div>
    <div
      class="c3"
    >
      <button
        class="c4"
      >
        socialAccounts.linkNewAccount
      </button>
    </div>
  </div>
</div>
`;
